pkg/service/orchestrator/rerecord.go

```diff
@@ -0,0 +1,252 @@
+//go:build linux
+
+package orchestrator
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"sort"
+	"time"
+
+	"go.keploy.io/server/v2/pkg"
+	"go.keploy.io/server/v2/utils"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+func (o *Orchestrator) ReRecord(ctx context.Context) error {
+	// creating error group to manage proper shutdown of all the go routines and to propagate the error to the caller
+
+	var stopReason string
+	var err error
+
+	defer func() {
+		select {
+		case <-ctx.Done():
+		default:
+			err := utils.Stop(o.logger, stopReason)
+			if err != nil {
+				utils.LogError(o.logger, err, "failed to stop recording")
+			}
+		}
+	}()
+
+	// Get all the testsets
+	testSets, err := o.replay.GetAllTestSetIDs(ctx)
+	if err != nil {
+		errMsg := "Failed to get all testset IDs"
+		utils.LogError(o.logger, err, errMsg)
+		return err
+	}
+
+	// Sort the testsets to ensure that the testcases are re-recorded in the same order
+	sort.SliceStable(testSets, func(i, j int) bool {
+		return testSets[i] < testSets[j]
+	})
+
+	var SelectedTests []string
+
+	for _, testSet := range testSets {
+		if ctx.Err() != nil {
+			break
+		}
+
+		if _, ok := o.config.Test.SelectedTests[testSet]; !ok && len(o.config.Test.SelectedTests) != 0 {
+			continue
+		}
+
+		SelectedTests = append(SelectedTests, testSet)
+
+		o.logger.Info("Re-recording testcases for the given testset", zap.String("testset", testSet))
+
+		errGrp, _ := errgroup.WithContext(ctx)
+		recordCtx := context.WithoutCancel(ctx)
+		recordCtx, recordCtxCancel := context.WithCancel(recordCtx)
+
+		var errCh = make(chan error, 1)
+		var replayErrCh = make(chan error, 1)
+
+		select {
```

Comment:
why do we have 2 back-2-back selects?

Reply:
To avoid calling `o.replayTests(recordCtx, testSet)` if the ctx is done just after the first go routine. Since  `replayTests` is a blocking call.

Reply:
Here we've used child context without cancel to avoid the cancellation of the parent context. When we use errgroup and get an error from any of the go routines spawned by errgroup, it cancels the parent context. We don't want to stop the execution if there is an error in any of the test-set recording sessions, it should just skip that test-set and continue with the next one.

Keeping two back-to-back selects is used to not do blocking operation if parent ctx is done